### ISSUE TITLE
Parse "Match Exec" in ssh config file

### ISSUE THREAD

**Issue by jarrah31** [satiromarra/vscode-sftp#21]
**Do you read the FAQ?**
- [x] Yes.

**Describe the bug**
Do you know if SFTP has the ability to parse "Match exec" within .ssh/config please?
If I specify a host such as "dsrv2" from the config below, I receive an error that says "Error: [dsrv2]: getaddrinfo ENOTFOUND dsrv2"
```
Match exec "/Users/me/onsubnet 192.168.1." host dsrv2
 Hostname 192.168.1.32
 Port 22
 User server

Match exec "/Users/me/onsubnet --not 192.168.1." host dsrv2
 ProxyJump mypi
 Hostname 192.168.1.32
 User server
 Port 22

Host mypi
 HostName my.domain.com
 Port 22
 User pi
```
onsubnet is a bash script that returns a 0 or 1 depending on which network my machine is on. It works great with Sublime's SFTP plugin but it's now the final thing holding me back from using vscode.

onsubnet script:
```
#!/usr/bin/env bash

if [[ "$1" == "--help" ]] || [[ "$1" == "-h" ]]  || [[ "$1" == "" ]] ; then
  printf "Usage:\n\tonsubnet [ --not ] partial-ip-address\n\n"
  printf "Example:\n\tonsubnet 10.10.\n\tonsubnet --not 192.168.0.\n\n"
  printf "Note:\n\tThe partial-ip-address must match starting at the first\n"
  printf "\tcharacter of the ip-address, therefore the first example\n"
  printf "\tabove will match 10.10.10.1 but not 110.10.10.1\n"
  exit 0
fi

on=0
off=1
if [[ "$1" == "--not" ]] ; then
  shift
  on=1
  off=0
fi

regexp="^$(sed 's/\./\\./g' <<<"$1")"

if [[ "$(uname)" == "Darwin" ]] ; then
  ifconfig | fgrep 'inet ' | fgrep -v 127.0.0. | cut -d ' ' -f 2 | egrep "$regexp" >/dev/null
else
  hostname -I | tr -s " " "\012" | fgrep -v 127.0.0. | egrep "$regexp" >/dev/null
fi

if [[ $? == 0 ]]; then
  exit $on
else
  exit $off
fi
```
If I specify a normal .ssh/config host, e.g. as shown below with a local IP, it resolves the hostname ok, so this at least indicates it's using the config file ok.
```
Host dsrvtest
 Hostname 192.168.1.32
 Port 22
 User server
```

**To Reproduce**
1. Use the .ssh/config settings and onsubnet script shown above.
2. I have the following sftp.json config:
```
{
    "name": "Docker Server 2",
    "host": "dsrv2",
    "protocol": "sftp",
    "port": 22,
    "username": "server",
    "remotePath": "/home/server/docker",
    "uploadOnSave": true,
    "useTempFile": false,
    "openSsh": false,
    "agent": "/Users/me/Library/Group Containers/XXXXXXX.com.1password/t/agent.sock",
    "sshConfigPath": "/Users/me/.ssh/config"
}
```

**Expected behavior**
It should connect to 192.168.1.32 if my laptop has an IP on my home network (192.168.1.x), or connect via the ProxyJump Pi using my external domain name.

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: [MacOS 13]
 - VSCode Version [1.76.2]
 - Extension Version [e.g. 1.17.9]

**Extension Logs from Startup** - *required*
  1.  Open User Settings.

      * On Windows/Linux - File > Preferences > Settings
      * On macOS - Code > Preferences > Settings
  4. Set `sftp.debug` to `true` and reload vscode.
  5. Reproduce the problem, get the logs from View > Output > sftp.

```
[2023-03-24 16:07:23] [debug] Custom crypto binding not available
[2023-03-24 16:07:23] [debug] Local ident: 'SSH-2.0-ssh2js1.11.0'
[2023-03-24 16:07:23] [debug] Client: Trying dsrv2 on port 22 ...
[2023-03-24 16:07:23] [debug] Socket error: getaddrinfo ENOTFOUND dsrv2
[2023-03-24 16:07:23] [error] Error: [dsrv2]: getaddrinfo ENOTFOUND dsrv2
	at Client.<anonymous> (/Users/me/.vscode/extensions/satiromarra.code-sftp-1.17.9/dist/extension.js:2:291992)
	at Client.emit (node:events:538:35)
	at Socket.<anonymous> (/Users/me/.vscode/extensions/satiromarra.code-sftp-1.17.9/node_modules/ssh2/lib/client.js:745:12)
	at Socket.emit (node:events:526:28)
	at emitErrorNT (node:internal/streams/destroy:157:8)
	at emitErrorCloseNT (node:internal/streams/destroy:122:3)
	at process.processTicksAndRejections (node:internal/process/task_queues:83:21) download /Users/me/Library/Mobile Documents/com~apple~CloudDocs/configuration.yaml
[2023-03-24 16:07:23] [debug] Socket closed
```

Thanks in advance for your help.

**Comment by satiromarra** [comment]
@jarrah31 You can try this beta?  [1.17.10-beta2](https://github.com/satiromarra/vscode-sftp/releases/download/v1.17.10-beta2/code-sftp-1.17.10-beta2.vsix)

**Comment by jarrah31** [comment]
> @jarrah31 You can try this beta?  [1.17.10-beta2](https://github.com/satiromarra/vscode-sftp/releases/download/v1.17.10-beta2/code-sftp-1.17.10-beta2.vsix)

Hi @satiromarra, I'm very excited to try the beta version but as it's 11pm here in the UK i'll unfortunately have to wait till tomorrow to try it out. I so very much appreciate you taking the time to help and can't thank you enough for having the capability to improve and build upon this excellent SFTP extension. 👍 👍

**Comment by jarrah31** [comment]
Hi @satiromarra 

Testing has resulted in a partial success (thank you!). 

If I'm on my home network, SFTP finds the Match entry and connects to 192.168.1.32:
```
[2023-03-25 13:35:05] [info] Section for 'dsrv2' not found with Host, try Match
[2023-03-25 13:35:05] [info] Section for 'dsrv2' found with Host or Match
[2023-03-25 13:35:05] [info] [file-save] /Users/me/Library/Mobile Documents/com~apple~CloudDocs/pkg_climate.yaml
[2023-03-25 13:35:05] [info] Section for 'dsrv2' not found with Host, try Match
[2023-03-25 13:35:06] [info] Section for 'dsrv2' found with Host or Match
[2023-03-25 13:35:06] [info] local ➞ remote /Users/me/Library/Mobile Documents/com~apple~CloudDocs/pkg_climate.yaml
```
However if I connect my laptop via my phone's hotspot for an external connection, it still tries to connect to my local IP. I suspect it isn't executing the onsubnet script to determine which host to connect to. 
```
[2023-03-25 13:47:37] [info] Section for 'dsrv2' not found with Host, try Match
[2023-03-25 13:47:38] [info] Section for 'dsrv2' found with Host or Match
[2023-03-25 13:47:38] [info] [file-save] /Users/me/Library/Mobile Documents/com~apple~CloudDocs/pkg_climate.yaml
[2023-03-25 13:47:38] [info] Section for 'dsrv2' not found with Host, try Match
[2023-03-25 13:47:38] [info] Section for 'dsrv2' found with Host or Match
[2023-03-25 13:47:48] [error] Error: [192.168.1.32]: Timed out while waiting for handshake
	at Client.<anonymous> (/Users/me/.vscode/extensions/satiromarra.code-sftp-1.17.10-beta2/dist/extension.js:2:290183)
	at Client.emit (node:events:538:35)
	at Timeout._onTimeout (/Users/me/.vscode/extensions/satiromarra.code-sftp-1.17.10-beta2/node_modules/ssh2/lib/client.js:1016:16)
	at listOnTimeout (node:internal/timers:559:17)
	at process.processTimers (node:internal/timers:502:7) download /Users/me/Library/Mobile Documents/com~apple~CloudDocs/pkg_climate.yaml
```
I modified my config file to remove ProxyJump for now (I suspect that is an added complication to be looked at once this works) so that it looks like this:
```
Match exec "/Users/me/onsubnet 192.168.1." host dsrv2
 Hostname 192.168.1.32
 Port 22
 User server

Match exec "/Users/me/onsubnet --not 192.168.1." host dsrv2
 HostName my.domain.com
 Port 22
 User pi
```
To show what the onsubnet script does, I can see that the --not option responds with a '0' success when connected externally.
```
me@MacBook ➜  ~ ./onsubnet --not 192.168.1.
me@MacBook ➜  ~ echo $?
0
me@MacBook ➜  ~ ./onsubnet 192.168.1.
me@MacBook ➜  ~ echo $?
1
```
This is the result when on my local network:
```
me@MacBook ➜  ~ ./onsubnet --not 192.168.1.
me@MacBook ➜  ~ echo $?
1
me@MacBook ➜  ~ ./onsubnet 192.168.1.
me@MacBook ➜  ~ echo $?
0
```
Hope this helps and thanks for your time.

**Comment by satiromarra** [comment]
@jarrah31 The "proxyjump" directive is complicated to implement. For now I can leave the search for the "Match Exec" directive that works well. right?

**Comment by jarrah31** [comment]
Yes, I agree that proxyjump would be tricky to implement so happy to leave that for now.

Unfortunately "Match Exec" doesn't quite work right yet, and it's this problem that I hopefully tried to explain in my previous post. 

Just out of interest, does the SFTP plugin use the hosts built-in sftp client (/usr/bin/sftp) when running on Linux/MacOS? If not do you think this would be possible as it should automatically parse .ssh/config and save you having to manually parse it? Maybe it could be a user configurable option so that SFTP still works on hosts (such as Windows) without the sftp client?

I realise this may be going way beyond what the SFTP plugin is capable of and would understand if it's too difficult to implement.

**Comment by satiromarra** [comment]
Is the latest update still failing you? I have tried the same "ontest" script and similar ssh config and it works correctly for me.

The implementation of a new client or allowing the user to modify it, it may need a refactoring of the base/core of the extension.  I don't rule it out but I don't think it can be done now.

**Comment by jarrah31** [comment]
> Is the latest update still failing you? I have tried the same "ontest" script and similar ssh config and it works correctly for me.
> 
> 
> 

Ah my apologies, i'll try again in the morning, I may have been doing something wrong. Just to check, is it the beta2 version I should try?


> The implementation of a new client or allowing the user to modify it, it may need a refactoring of the base/core of the extension.  I don't rule it out but I don't think it can be done now.

No problems, thank you for considering it.

**Comment by jarrah31** [comment]
Hi @satiromarra, it turns out I needed to fully quit VSC rather than just reload the plugin as it doesn't dynamically pick up on changes made within .ssh/config. It was still using the ProxyJump config which had my local IP as the Hostname, even though I had modified and saved the config file to use a domain name instead.
Is this behaviour expected or is it a bug? 

The other mistake I made was forgetting to remove the Port entry in sftp.json when I had a different port in .ssh/config, so that had me puzzled for a few mins. :) 

Other than that it seems to work well! I need to do some more testing to figure out a way around ProxyJump, but for now it's working great.

Thank you so much for adding this feature!
